### ISSUE TITLE
Uninitialized variable used in expression

### ISSUE THREAD

**Issue by pkomiske** [tensor-compiler/taco#134]
With taco's "Try online" feature, if I enter the expression:
    `result = a(i1)*a(i2)*a(i3)*X(i1,i2)*X(i1,i3)`
(all as dense tensors) then it uses a variable `pa1` in "Compute Loops" (or `pa11` in "Complete Code") without initializing its value. When I try to evaluate this expression in a custom program, I get 0 (which is wrong). 

If this is somehow an unsupported type of expression, then it needs to be documented as such somewhere.

**Comment by fredrikbk** [comment]
@pkomiske, @stephenchouca and I met to discuss this. It looks like this is a bug with code generation and I'll look at it this week and next.

@pkomiske also sent us some test cases:

"Here are two files to help you in the debugging. `events.txt` contains ten random events (which each have a vector z and a matrix theta). The dimension M is 25 for all these events.  I’ve flattened the matrix (and all other higher-order tensors that appear) in row-major order.

`expressions.txt` contains 156 expressions returning a scalar, 5 expressions returning a vector, 6 expressions returning a matrix, and 3 expressions returning rank-3 tensors. The format of this file is a commented line (starting with ‘#’) indicating a unique index of that expression, then the expression, and then the value of that expression for each of the ten events, one per line."

[events.txt](https://github.com/tensor-compiler/taco/files/1643526/events.txt)
[expressions.txt](https://github.com/tensor-compiler/taco/files/1643527/expressions.txt)

**Comment by fredrikbk** [comment]
The expression from op (`tau = z(i) * z(j) * z(k) * theta(i,j) * theta(i,k)`) emits as:
```c
tau_vals[0] = 0;
for (int32_t iz = 0; iz < z1_size; iz++) {
  double tj = 0;
  for (int32_t jz = 0; jz < z1_size; jz++) {
    int32_t ptheta2 = (iz * theta2_size) + jz;
    tj += ((z_vals[iz] * z_vals[jz]) * z_vals[pz11]) * theta_vals[ptheta2];
  }
  double tk = 0;
  for (int32_t ktheta = 0; ktheta < theta2_size; ktheta++) {
    int32_t ptheta20 = (iz * theta2_size) + ktheta;
    tk += theta_vals[ptheta20];
  }
  tau_vals[0] = tau_vals[0] + (tj * tk);
}
```
The problem is that the `z(k)` sub-expression is emitted in the i,j loop, indexed by the a z position variable. It should be emitted in the i,k loop:

```c
tau_vals[0] = 0;
for (int32_t iz = 0; iz < z1_dimension; iz++) {
  double tj = 0;
  for (int32_t jz = 0; jz < z1_dimension; jz++) {
    int32_t ptheta2 = (iz * theta2_dimension) + jz;
    tj += z_vals[iz] * z_vals[jz] * theta_vals[ptheta2];
  }
  double tk = 0;
  for (int32_t ktheta = 0; ktheta < theta2_dimension; ktheta++) {
    int32_t ptheta20 = (iz * theta2_dimension) + ktheta;
    tk += z_vals[ktheta] * theta_vals[ptheta20];
  }
  tau_vals[0] = tau_vals[0] + (tj * tk);
}
```

**Comment by fredrikbk** [comment]
Test case ('expr-qcd-tests.cpp`):

```c
TEST(expr, qcd3) {
  Tensor<double> tau("tau");
  Tensor<double> z = read(qcdTestData("z0.ttx"), Dense);
  Tensor<double> theta = read(qcdTestData("theta0.ttx"), Dense);

  IndexVar i, j, k;
  tau = z(i) * z(j) * z(k) * theta(i,j) * theta(i,k);

  tau.evaluate();
  ASSERT_DOUBLE_EQ(0.3223971010027145, getScalarValue(tau));
}
```

**Comment by fredrikbk** [comment]
This is fixed with fa8f686555b3d9862aae57c0004e1e0f2623fdcd.

@pkomiske let us know if anything else does not work for you!

The solution is temporary and quite hacky, but I think it's fine for a little bit. The investigation into this bug exposed a need to redesign our system to include summation nodes in the index expression IR, nd the temporary solution is just there while I do this.
